src/expr/window.rs

```diff
@@ -0,0 +1,297 @@
+                }
+            },
+        };
+        let end_bound = match end_bound {
+            Some(end_bound) => WindowFrameBound::Following(ScalarValue::UInt64(Some(end_bound))),
+            None => match units {
+                WindowFrameUnits::Rows => WindowFrameBound::Following(ScalarValue::UInt64(None)),
+                WindowFrameUnits::Range => WindowFrameBound::Following(ScalarValue::UInt64(None)),
+                WindowFrameUnits::Groups => {
+                    return Err(py_datafusion_err(DataFusionError::NotImplemented(format!(
+                        "{:?}",
+                        units,
+                    ))));
+                }
+            },
+        };
+        Ok(PyWindowFrame {
+            window_frame: WindowFrame {
+                units,
+                start_bound,
+                end_bound,
+            },
+        })
+    }
+
+    /// Returns the window frame units for the bounds
+    pub fn get_frame_units(&self) -> PyResult<String> {
+        Ok(self.window_frame.units.to_string())
+    }
+    /// Returns starting bound
+    pub fn get_lower_bound(&self) -> PyResult<PyWindowFrameBound> {
+        Ok(self.window_frame.start_bound.clone().into())
+    }
+    /// Returns end bound
+    pub fn get_upper_bound(&self) -> PyResult<PyWindowFrameBound> {
+        Ok(self.window_frame.end_bound.clone().into())
+    }
+
+    /// Get a String representation of this window frame
+    fn __repr__(&self) -> String {
+        format!("{}", self)
+    }
+}
+
+#[pymethods]
+impl PyWindowFrameBound {
+    /// Returns if the frame bound is current row
+    pub fn is_current_row(&self) -> bool {
+        matches!(self.frame_bound, WindowFrameBound::CurrentRow)
+    }
+
+    /// Returns if the frame bound is preceding
+    pub fn is_preceding(&self) -> bool {
+        matches!(self.frame_bound, WindowFrameBound::Preceding(_))
+    }
+
+    /// Returns if the frame bound is following
+    pub fn is_following(&self) -> bool {
+        matches!(self.frame_bound, WindowFrameBound::Following(_))
+    }
+    /// Returns the offset of the window frame
+    pub fn get_offset(&self) -> PyResult<Option<u64>> {
+        match &self.frame_bound {
+            WindowFrameBound::Preceding(val) | WindowFrameBound::Following(val) => match val {
+                x if x.is_null() => Ok(None),
+                ScalarValue::UInt64(v) => Ok(*v),
+                // The cast below is only safe because window bounds cannot be negative
+                ScalarValue::Int64(v) => Ok(v.map(|n| n as u64)),
+                ScalarValue::Utf8(v) => {
+                    let s = v.clone().unwrap();
+                    match s.parse::<u64>() {
+                        Ok(s) => Ok(Some(s)),
+                        Err(_e) => Err(DataFusionError::Plan(format!(
+                            "Unable to parse u64 from Utf8 value '{s}'"
+                        ))
+                        .into()),
+                    }
+                }
```

Comment:
nit: we can avoid the `unwrap` here:

```suggestion
                ScalarValue::Utf8(Some(s)) => match s.parse::<u64>() {
                    Ok(s) => Ok(Some(s)),
                    Err(_e) => Err(DataFusionError::Plan(format!(
                        "Unable to parse u64 from Utf8 value '{s}'"
                    ))
                    .into()),
                },
```